NIFI-8516 Enabled HTTPS and Single User Authentication by default

exceptionfactory opened the pull request:
#### Description of PR

NIFI-8516 Implements the stated goals of NIFI-8220 providing a basic layer of security for the default installation of NiFi.

The default configuration disables HTTP on port 8080 and enables HTTPS on port 9443 using a self-signed certificate generated during the bootstrap process.  The default configuration also leverages the Single User Login Identity Provider and Authorizer to generate a random username and password for authentication.  The NiFi README.md includes updated instructions and screenshots describing how to access NiFi using the generated credentials.

These changes will not impact existing installations of NiFi following an upgrade.  Neither the self-signed certificate nor the generated username and password are intended for production deployments.  The purpose of these default configuration changes is to provide a greater level of security for default installations of NiFi while maintain a straightforward experience for new users.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [X] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

gresockj commented on nifi-docs/src/main/asciidoc/administration-guide.adoc:
```diff
@@ -3527,7 +3527,7 @@ For example, to provide two additional network interfaces, a user could also spe
  +
 Providing three total network interfaces, including  `nifi.web.http.network.interface.default`.
 |`nifi.web.https.host`|The HTTPS host. It is blank by default.
```
Should change the language for nifi.web.https.host since it's no longer blank by default.  Reverse comment for nifi.web.http.host.

exceptionfactory replied:
Thanks for catching that detail, I pushed an update with changes.

gresockj reviewed (changes requested):
Great work, this has a nice feel to it when you put it all together!  Just documentation comments below.  Also, we should update the Getting Started guide, which still references localhost:8080.

exceptionfactory commented:
> Great work, this has a nice feel to it when you put it all together! Just documentation comments below. Also, we should update the Getting Started guide, which still references localhost:8080.

Thanks for the feedback @gresockj!  I pushed an update with additional details in the Getting Started guide describing the generated credentials and referencing to new default URL.

gresockj reviewed (approved):
LGTM!

ChrisSamo632 commented:
Haven't tried it to confirm, but I don't think this will work with the Docker start scripts, which still default to no authentication (or is that a conscious decision)?

exceptionfactory commented:
> Haven't tried it to confirm, but I don't think this will work with the Docker start scripts, which still default to no authentication (or is that a conscious decision)?

Thanks for the feedback @ChrisSamo632, that's a good question.  The change to the default properties will impact the Docker image.  It looks like some adjustments to the startup property configuration in the Docker image is going to be necessary.  I will look into it and evaluate potential changes.

thenatog commented:
Will review

thenatog commented:
Tested this out and seems to work well. One question would be, why specifically is the default port 9443? Typically, most documentation refers to NiFi using 8443 for secure NiFi:
https://pierrevillard.com/2016/11/29/apache-nifi-1-1-0-secured-cluster-setup/
https://community.cloudera.com/t5/Community-Articles/Securing-NiFi-Step-by-Step/ta-p/244315
https://community.cloudera.com/t5/Support-Questions/NiFi-Cannot-replicate-request-to-Node-nifi-domain-net-8443/td-p/283664

exceptionfactory commented:
> Tested this out and seems to work well. One question would be, why specifically is the default port 9443? Typically, most documentation refers to NiFi using 8443 for secure NiFi:
> https://pierrevillard.com/2016/11/29/apache-nifi-1-1-0-secured-cluster-setup/
> https://community.cloudera.com/t5/Community-Articles/Securing-NiFi-Step-by-Step/ta-p/244315
> https://community.cloudera.com/t5/Support-Questions/NiFi-Cannot-replicate-request-to-Node-nifi-domain-net-8443/td-p/283664

Thanks for the testing and feedback @thenatog!  The Administration Guide had a note about 9443 being the default for HTTPS, but this seems like a good opportunity to bring that in line with other common examples, so I will update the default value and documentation to use `8443`.

I am in the process of making some adjustments to the Docker scripts to enable HTTPS by default, so I will push these updates soon.

thenatog commented:
Hmm, yes, it does appear to be a bit of a mix between 8443 and 9443. Perhaps we should continue with 9443 as that's what officially existed in docs.

exceptionfactory commented:
@ChrisSamo632, I updated the Docker startup scripts to enable HTTPS on port 8443 as the default configuration.

@thenatog In light of the Docker images previously referencing 8443 as the default port for HTTPS, and with 8443 being more common as a default port for HTTPS, I updated also updated the default value for `nifi.properties` as well as the associated getting started documentation.

Please feel free to provide any additional feedback.

ChrisSamo632 commented on nifi-docker/dockerhub/sh/start.sh:
```diff
@@ -33,15 +33,29 @@ if [ ! -z "${NIFI_JVM_DEBUGGER}" ]; then
-prop_replace 'baseUrl' "http://${NIFI_WEB_HTTP_HOST:-$HOSTNAME}:${NIFI_WEB_HTTP_PORT:-8080}" ${nifi_toolkit_props_file}
+prop_replace 'baseUrl' "https://${NIFI_WEB_HTTPS_HOST:-$HOSTNAME}:${NIFI_WEB_HTTPS_PORT:-8443}" ${nifi_toolkit_props_file}
+
+if [ ! -z "${NIFI_WEB_HTTP_PORT}" ]; then
```
Could use `-n` instead of `! -z` to test whether the env var is populated

exceptionfactory replied:
Thanks, will make the adjustment.

ChrisSamo632 replied:
Changes look good (I can't resolve this as a thread for some reason)

ChrisSamo632 commented on nifi-docker/dockermaven/sh/start.sh:
```diff
@@ -33,15 +33,29 @@ if [ ! -z "${NIFI_JVM_DEBUGGER}" ]; then
+    prop_replace 'nifi.web.https.host'                        ''
+    prop_replace 'nifi.web.http.port'                         "${NIFI_WEB_HTTP_PORT}"
+    prop_replace 'nifi.web.http.host'                         "${NIFI_WEB_HTTP_HOST:-$HOSTNAME}"
+    prop_replace 'nifi.remote.input.secure'                   'false'
```
Should also set `nifi.cluster.protocol.is.secure` (I think this is a historical omission but makes sense to correct)

These comments apply to both sets of the docker start scripts

exceptionfactory replied:
Yes, this is a good opportunity to make the adjustment.

ChrisSamo632 replied:
Changes look good (I can't resolve this as a thread for some reason)

exceptionfactory replied:
Thanks again @ChrisSamo632, I appreciate the thorough review!

thenatog commented:
On further assessment, lets use port 8443 rather than 9443.

thenatog commented on nifi-docker/dockerhub/README.md:
```diff
@@ -62,29 +66,29 @@ The configuration scripts are suitable for at least 1.4.0+.
 
     docker run --name nifi \
-      -p 9090:9090 \
+      -p 9443:9443 \
```
Do we want this now to be 8443?

ChrisSamo632 replied:
This is an example of overriding the default (8443), I think this is fine

exceptionfactory replied:
Thanks @ChrisSamo632, I agree, this is part of the example showing an alternative port configuration.

markap14 commented:
Thanks @exceptionfactory. This appears to work as designed. However, I think that this is a pretty concerning change to the user experience. Every time the user wants to login, they now are required to go find that generated username and password. There appears to be no way to set your own. I think we need the ability to set the username & password somehow. Preferably with something like `nifi.sh set-single-user <user> <pass>` or something like that.

Would also recommend using a pre-defined username such as `admin` or `nifi` or something but I can leave that up to your discretion if the commandline util gives us the ability to set username and password

exceptionfactory commented:
> Thanks @exceptionfactory. This appears to work as designed. However, I think that this is a pretty concerning change to the user experience. Every time the user wants to login, they now are required to go find that generated username and password. There appears to be no way to set your own. I think we need the ability to set the username & password somehow. Preferably with something like `nifi.sh set-single-user <user> <pass>` or something like that.
> 
> Would also recommend using a pre-defined username such as `admin` or `nifi` or something but I can leave that up to your discretion if the commandline util gives us the ability to set username and password

Thanks for the review and feedback @markap14!  Although Single User authentication is not intended for production use cases, having the ability to change the credentials would certainly make it easier to get started.

I pushed an update that includes a new command option:

`nifi.sh set-single-user-credentials <username> <password>`

This required refactoring shared classes to a new `nifi-single-user-utils` module that could be leveraged in the bootstrap library.  I also updated the Getting Started documentation with an example command, and included a log statement with the command after generating the default random credentials.

With this option, I kept the random username as well as random password.

Please let me know if you have any additional feedback.

gresockj commented on nifi-toolkit/nifi-toolkit-assembly/docker/Dockerfile:
```diff
@@ -43,7 +43,7 @@ RUN chown -R nifi:nifi ${NIFI_TOOLKIT_BASE_DIR}
 
 # Default port for TLS Toolkit CA Server
-EXPOSE 8443
+EXPOSE 9443
```
The toolkit guide adoc also needs to be updated for 9443.

exceptionfactory replied:
Thanks, good catch.

gresockj reviewed (changes requested):
Functions as designed, just a couple documentation changes needed.

exceptionfactory commented on README.md:
```diff
@@ -65,67 +65,108 @@ Apache NiFi was made for dataflow. It supports highly configurable directed grap
+
+The `USERNAME` will be a random UUID composed of 36 characters. The `PASSWORD` will be a random string composed of
+32 characters. The generated credentials will be stored in `conf/login-identity-providers.xml` with the password stored
+using bcrypt hashing. Record these credentials in a secure location for access to NiFi.
```
Yes, will add an example of the command.

exceptionfactory commented on nifi-docs/src/main/asciidoc/administration-guide.adoc:
```diff
@@ -84,15 +84,14 @@ The following table lists the default ports used by NiFi and the corresponding p
-NOTE: The ports marked with an asterisk (*) have property values that are blank by default in _nifi.properties_.  The values shown in the table are the default values for these ports when <<toolkit-guide.adoc#tls_toolkit,TLS Toolkit>> is used to generate _nifi.properties_ for a secured NiFi instance.  The default Certificate Authority Port used by <<toolkit-guide.adoc#tls_toolkit,TLS Toolkit>> is ...
+NOTE: The ports marked with an asterisk (*) have property values that are blank by default in _nifi.properties_.  The values shown in the table are the default values for these ports when <<toolkit-guide.adoc#tls_toolkit,TLS Toolkit>> is used to generate _nifi.properties_ for a secured NiFi instance.  The default Certificate Authority Port used by <<toolkit-guide.adoc#tls_toolkit,TLS Toolkit>> is ...
 
 === Embedded ZooKeeper
```
Thanks, I updated that Administration Guide section and included a new section for Single User authentication.

exceptionfactory commented:
> Functions as designed, just a couple documentation changes needed.

Thanks for the additional feedback @gresockj!

thenatog commented:
So was just doing a bit of testing, and I found that if you run ./bin/nifi.sh set-single-user-credentials <username> <password> with a different secure configuration (eg LDAP), the password/user will not be set. This is good, however it doesn't necessarily indicate that the command had no effect, it will still state "Login Identity Providers Processed".

exceptionfactory commented:
> So was just doing a bit of testing, and I found that if you run ./bin/nifi.sh set-single-user-credentials with a different secure configuration (eg LDAP), the password/user will not be set. This is good, however it doesn't necessarily indicate that the command had no effect, it will still state "Login Identity Providers Processed". Otherwise, things are looking good.

Thanks for pointing that out @thenatog.  Do you think any changes are necessary in that case?  It seems that the command name `set-single-user-credentials` indicates what the command does, so unless the command should throw an error, I'm not sure what else should happen.

thenatog commented:
I would maybe expect to see a log saying "Single user provider cannot be used in conjunction with other identity providers."?

exceptionfactory commented:
> I would maybe expect to see a log saying "Single user provider cannot be used in conjunction with other identity providers."?

That could be possible by evaluating the configured value from `nifi.properties` against the configured providers in login-identity-providers.xml, but it would involve some amount of additional handling just to produce an error message.  If you think it is worth doing, I can make the changes.

thenatog commented:
Okay no problem, I don't think it's worth it.

markap14 commented:
Thanks for the update @exceptionfactory! Things are looking really good here! I did some testing and playing around with it and all works well. I'll do a code review and if all looks good merge.

Only thing I noticed was when I tried to set the password to something short like "pass" I got the following output:
```
Java home: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home
NiFi home: /Users/mpayne/devel/nifi/nifi-assembly/target/nifi-1.14.0-SNAPSHOT-bin/nifi-1.14.0-SNAPSHOT

Bootstrap Config File: /Users/mpayne/devel/nifi/nifi-assembly/target/nifi-1.14.0-SNAPSHOT-bin/nifi-1.14.0-SNAPSHOT/conf/bootstrap.conf

Password length less than required [12]
```
In that case, we have basically some debug output and then a failure message, but the failure message IMO wasn't as clear that it could be. It would be easy to miss the fact that it was a failure, and that could lead to a lot of confusion. So if all else looks good I'll update that message on merge to indicate "ERROR: " as a prefix so it's obvious. Otherwise, all looking great!

markap14 commented:
All looks good @exceptionfactory. I've merged to main! Many thanks to @exceptionfactory for implementing this and to @thenatog @gresockj and @ChrisSamo632 for reviewing!
